sdk/include/ds/linked_list.h

```diff
@@ -0,0 +1,642 @@
+	}
+	/** @} */
+
+	/**
+	 * Link equality predicates
+	 *
+	 * @{
+	 */
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<!cell::HasEqOps<T>, bool> is_prev_eq(T *e,
+	                                                                      T *p)
+	{
+		return e->cell_prev() == p;
+	}
+
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<cell::HasEqOps<T>, bool> is_prev_eq(T *e,
+	                                                                     T *p)
+	{
+		return e->cell_is_prev_eq(p);
+	}
+
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<!cell::HasEqOps<T>, bool> is_next_eq(T *e,
+	                                                                      T *p)
+	{
+		return e->cell_next() == p;
+	}
+
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<cell::HasEqOps<T>, bool> is_next_eq(T *e,
+	                                                                     T *p)
+	{
+		return e->cell_is_next_eq(p);
+	}
+	/** @} */
+
+	/**
+	 * Verify linkage invariants.  Again, useful only for testing.
+	 *
+	 * The default implementation decodes all four relevant links.
+	 */
+	template<cell::HasPrimOps T>
+	__always_inline bool is_well_formed(T *e)
+	{
+		return (e == e->cell_prev()->cell_next()) &&
+		       (e == e->cell_next()->cell_prev());
+	}
+
+	/**
+	 * Insert a ring of `elem`-ents (typically, a singleton ring) before the
+	 * `curr`-ent element (or sentinel) in the ring.  In general, you will
+	 * probably want to make sure that at most one of `elem` or `curr`
+	 * points to a ring with a sentinel node.
+	 *
+	 * If `curr` is the sentinel, this is appending to the list, in the
+	 * sense that the element(s) occupy (or span) the next-most and
+	 * prev-least position from the sentinel.
+	 *
+	 * By symmetry, if `elem` is, instead, the sentinel, then `curr` is
+	 * prepended to the list in the same sense.
+	 */
+	template<cell::HasPrimOps Cell>
+	__always_inline void insert_before(Cell *curr, Cell *elem)
+	{
+		curr->cell_prev()->cell_next() = elem->cell_next();
+		elem->cell_next()->cell_prev() = curr->cell_prev();
+		curr->cell_prev()              = elem;
+		elem->cell_next()              = curr;
+	}
+
+	/**
+	 * Fuse cell initialization with insertion before.  Specifically,
+	 *
+	 *    insert_new_before(c, e);
+	 *
+	 * is semantically equivalent to
+	 *
+	 *    e->cell_init(); insert_before(c, e);
+	 *
+	 * but spelled in a way that the compiler can understand a bit better, with
+	 * less effort spent in provenance and/or alias analysis.
+	 */
+	template<cell::HasPrimOps Cell, typename P>
+	requires std::same_as<P, Cell *> || ds::pointer::proxy::Proxies<P, Cell>
+	  __always_inline void insert_new_before(P curr, Cell *elem)
+	{
+		auto prev         = curr->cell_prev();
+		elem->cell_next() = curr;
+		elem->cell_prev() = prev;
+		prev->cell_next() = elem;
+		prev              = elem;
+	}
+
+	/**
+	 * Fuse cell initialization with insertion after.  Specifically,
```

Comment:
Is this the same as `emplace` in standard-library terminology?

Reply:
Yes.  I've renamed things appropriately.